src/plugins/intel_npu/src/al/include/npu.hpp

```diff
@@ -81,8 +81,8 @@ class IDevice : public std::enable_shared_from_this<IDevice> {
     virtual std::map<ov::element::Type, float> getGops() const;
 
     virtual std::shared_ptr<SyncInferRequest> createInferRequest(
-        const std::shared_ptr<const ICompiledModel>& compiledModel,
-        const std::shared_ptr<IExecutor>& executor,
+        const std::shared_ptr<const ICompiledModel> compiledModel,
+        const std::shared_ptr<IExecutor> executor,
```

Comment:
Downside is countable here: counter increments upon entry, decrements on exit. 
Is verbal explanation of the achieved added value available?

Reply:
using std::move as well

Reply:
That was just one example of multiple instances of removed `const std::shared_ptr<>&` all over the files. I highlighted single only to keep the PR cleaner. 
Std-Moving the smart pointer clutters the code and will not be logically correct for every such change.

Reply:
If original version was `const std::shared_ptr<T>&`, it means upon call no ownership were moved or shared, changing to `const std::shared_ptr<T>` doesn't make API cleaner, but does add overhead to efficiency **and changes ownership semantics, and so basically breaks / changes the contract**

My opinion:

It's better to make API to express the intention as clear as possible: 

if you want to share ownership during the call - take `std::shared_ptr` by value (`const` doesn't make any sense here, as it's implementation detail whether you want to change argument value or not - outside of user's scope) - this way copy happens on user side and it's clear from the interface ownership is shared through copy of smart pointer

if you don't want to share ownership and just want to read the resource relying on calling context to keep resource alive - don't use smart pointers at all and take the type directly - `const T&` or `const T*` - raw pointer is perfectly fine here as it clearly signal you don't own resource without clatter and restrictions to pass smart pointer (remember there's nothing wrong with raw pointers even when smart ones arrived, the latter is a just a tool to make ownership strategy explicit when it's needed)

if you don't want to share, but want to overtake the ownership, use r-value reference `std::shared_ptr&&`, so that user would have to `std::move` upon call, again making ownership decision explicit

Reply:
Changed the code to use raw pointers as `T*` instead of passing smart pointers through the code.

Reply:
It feels like we have to start over with the value of lifecycles here.
[That](https://github.com/openvinotoolkit/openvino/pull/26307/commits/c78d37f0ca86cf6293ff1696d48543604f3b818e#diff-14b520c08aadc9f3c959e9173f246b6d7e7d6521e1b0acc1c3cdeafd7efd78acR79) is clear example of what could lead to program crash and therefore must be avoided. This line was introduced a bit earlier, but [is backed](https://github.com/openvinotoolkit/openvino/pull/26307/commits/c78d37f0ca86cf6293ff1696d48543604f3b818e#diff-92180db0acd07515dc319052f1bcb0a99a5df535a412252b971bdf1657ad01faL148) up with [smart ptr](https://github.com/openvinotoolkit/openvino/pull/26307/commits/c78d37f0ca86cf6293ff1696d48543604f3b818e#diff-14b520c08aadc9f3c959e9173f246b6d7e7d6521e1b0acc1c3cdeafd7efd78acL79) which would prevent the crash if the infer request object managed to live longer than the executor object and decided to use that stored raw pointer. Smart ptr would not save the program anymore with the last commit.

[When to Use Raw Pointers](https://hitchcock.codes/blog/when-to-use-raw-pointers) is relatively fresh and short article which does not deny using of raw pointers, briefly touches raw references and object lifecycles too. Found it just now.

Reply:
Prev stmt with links to the stable commits: 
[`ZeroExecutor* _executor`](https://github.com/openvinotoolkit/openvino/blob/fbe2b53c3c681348dbae10d7d3a3de0626902fbf/src/plugins/intel_npu/src/backend/include/zero_infer_request.hpp#L80) [is backed up with smart ptr](https://github.com/openvinotoolkit/openvino/blob/fbe2b53c3c681348dbae10d7d3a3de0626902fbf/src/plugins/intel_npu/src/backend/include/zero_infer_request.hpp#L79)